core/tauri-build/src/allowlist.rs

```diff
@@ -0,0 +1,213 @@
+}
+
+#[derive(Debug, Clone, Copy)]
+enum DependencyKind {
+  Build,
+  Normal,
+}
+
+#[derive(Debug)]
+struct AllowlistedDependency {
+  name: String,
+  alias: Option<String>,
+  kind: DependencyKind,
+  all_cli_managed_features: Option<Vec<&'static str>>,
+  expected_features: Vec<String>,
+}
+
+pub fn check(config: &Config, manifest: &mut Manifest) -> Result<()> {
+  let dependencies = vec![
+    AllowlistedDependency {
+      name: "tauri-build".into(),
+      alias: None,
+      kind: DependencyKind::Build,
+      all_cli_managed_features: Some(vec!["isolation"]),
+      expected_features: match config.tauri.pattern {
+        PatternKind::Isolation { .. } => vec!["isolation".to_string()],
+        _ => vec![],
+      },
+    },
+    AllowlistedDependency {
+      name: "tauri".into(),
+      alias: None,
+      kind: DependencyKind::Normal,
+      all_cli_managed_features: Some(TauriConfig::all_features()),
+      expected_features: config
+        .tauri
+        .features()
+        .into_iter()
+        .map(|f| f.to_string())
+        .collect::<Vec<String>>(),
+    },
+  ];
+
+  for metadata in dependencies {
+    let mut name = metadata.name.clone();
+    let mut dep = find_dependency(manifest, &metadata.name, metadata.kind);
+    if dep.is_none() {
+      if let Some(alias) = &metadata.alias {
+        dep = find_dependency(manifest, alias, metadata.kind);
+        name = alias.clone();
+      }
+    }
+
+    if let Some(dep) = dep {
+      let error_message = check_features(dep, &metadata);
+
+      if !error_message.is_empty() {
+        return Err(anyhow!("
+      The `{}` dependency features on the `Cargo.toml` file does not match the allowlist defined under `tauri.conf.json`.
+      Please run `tauri dev` or `tauri build` or {}.
+    ", name, error_message));
+      }
+    }
+  }
+
+  Ok(())
+}
+
+fn find_dependency(
+  manifest: &mut Manifest,
+  name: &str,
+  kind: DependencyKind,
+) -> Option<Dependency> {
+  let dep = match kind {
+    DependencyKind::Build => manifest.build_dependencies.remove(name),
+    DependencyKind::Normal => manifest.dependencies.remove(name),
+  };
+
+  if let Some(dep) = dep {
+    return Some(dep);
+  } else {
+    for target in manifest.target.values_mut() {
+      if let Some(dep) = match kind {
+        DependencyKind::Build => target.build_dependencies.remove(name),
+        DependencyKind::Normal => target.dependencies.remove(name),
+      } {
+        return Some(dep);
+      }
+    }
+  }
+
+  None
+}
+
+fn features_diff(current: &[String], expected: &[String]) -> Diff {
+  let mut remove = Vec::new();
+  let mut add = Vec::new();
+  for feature in current {
+    if !expected.contains(feature) {
+      remove.push(feature.clone());
+    }
+  }
+
+  for feature in expected {
+    if !current.contains(feature) {
+      add.push(feature.clone());
+    }
+  }
+
+  Diff { remove, add }
+}
+
+fn check_features(dependency: Dependency, metadata: &AllowlistedDependency) -> String {
```

Comment:
Should probably return result here and check for errors instead of empty string
```suggestion
fn check_features(dependency: Dependency, metadata: &AllowlistedDependency) -> Result<(), String > {
```